pydantic/dataclasses.py

```diff
@@ -54,6 +68,39 @@ def setattr_validate_assignment(self: 'DataclassType', name: str, value: Any) ->
     object.__setattr__(self, name, value)
 
 
+def is_builtin_dataclass(_cls: Type[Any]) -> bool:
+    """
+    `dataclasses.is_dataclass` is True if one of the class parents is a `dataclass`.
+    This is why we also add a class attribute `__processed__` to only consider 'direct' built-in dataclasses
+    """
+    return not hasattr(_cls, '__processed__') and dataclasses.is_dataclass(_cls)
+
+
+def _dataclass_with_validation(
```

Comment:
not sure it's worth this being a separate function?

Reply:
939f990